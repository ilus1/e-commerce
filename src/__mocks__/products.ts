export const mockProducts = [
  { id: '1', name: 'Produto 1', price: 10, photo: 'url1', description: 'Descrição 1' },
];

export const mockCartItems = [
  { id: '1', name: 'Produto 1', price: 10, photo: 'url1', description: 'Descrição 1', count: 1 },
];
