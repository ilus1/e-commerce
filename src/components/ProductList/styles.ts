import styled from 'styled-components'

export const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  flex-basis: 25%;
  gap: 22px;

  margin: 50px;
  width: 1000px;
`
