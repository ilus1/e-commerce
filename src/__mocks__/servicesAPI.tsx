import { rest } from "msw";
import { setupServer } from "msw/node";

import { mockProducts } from "./products";

const server = setupServer(
  rest.get('*', (_, res, ctx) => res(ctx.status(200), ctx.json(mockProducts)))
);

export default server;
