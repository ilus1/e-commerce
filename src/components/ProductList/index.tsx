import ProductComponent from '../../components/Product'
import { Container } from './styles'

interface Product {
  id: string
  name: string
  price: number
  photo: string
  description: string
}

interface CartItem {
  id: string
  name: string
  price: number
  photo: string
  description: string
  count: number
}

interface ProductListProps {
  products: Product[]
  cartItems: CartItem[]
  onAddProduct: (product: Product) => void
  onRemoveProduct: (product: CartItem) => void
}


function ProductList({ products, cartItems, onAddProduct, onRemoveProduct }: ProductListProps) {
  return (
    <Container data-testid="product-list">
      {products.map((product) => (
        <ProductComponent
          key={product.id}
          name={product.name}
          price={product.price}
          image={product.photo}
          description={product.description}
          count={cartItems.find((item) => item.id === product.id)?.count || 0}
          onAdd={() => onAddProduct(product)}
          onRemove={() => onRemoveProduct(cartItems.find((item) => item.id === product.id)!)}
        />
      ))}

    </Container>
  )
}

export default ProductList
