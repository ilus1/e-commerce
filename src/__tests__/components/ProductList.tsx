import renderer from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react';

import ProductList from '../../components/ProductList';

import { mockProducts, mockCartItems } from '../../__mocks__/products';

describe('Componente de Listagem de produtos', () => {
  it('Testa carregamento correto', () => {
    const mockAddProduct = jest.fn();
    const mockRemoveProduct = jest.fn();

    const tree = renderer
      .create(<ProductList
        isLoading={false}
        products={mockProducts}
        cartItems={mockCartItems}
        onAddProduct={mockAddProduct}
        onRemoveProduct={mockRemoveProduct}
      />)
      .toJSON()

    expect(tree).toMatchSnapshot();
  });

  it('Deve mostrar botao de subtrair items', () => {
    const cartItems = [{ ...mockCartItems[0], count: 2 }];
    const mockAddProduct = jest.fn();
    const mockRemoveProduct = jest.fn();

    render(<ProductList
      isLoading={false}
      products={mockProducts}
      cartItems={cartItems}
      onAddProduct={mockAddProduct}
      onRemoveProduct={mockRemoveProduct}
    />);

    const removeButton = screen.getByTestId('RemoveOutlinedIcon');
    fireEvent.click(removeButton);
    expect(mockRemoveProduct).toHaveBeenCalledTimes(1);
  });

  it('Deve mostrar botao de "comprar" produto ao carrinho', () => {
    const cartItems = [{ ...mockCartItems[0], count: 0 }];
    const mockAddProduct = jest.fn();
    const mockRemoveProduct = jest.fn();

    render(<ProductList
      isLoading={false}
      products={mockProducts}
      cartItems={cartItems}
      onAddProduct={mockAddProduct}
      onRemoveProduct={mockRemoveProduct}
    />);

    const button = screen.getByTestId('LocalMallOutlinedIcon').closest('button');
    expect(button?.textContent).toBe('Comprar');
    button?.click();
    expect(mockAddProduct).toHaveBeenCalledTimes(1);
  });

  it('Deve adicionar produto ao carrinho', () => {
    const cartItems = mockCartItems;
    const mockAddProduct = jest.fn(() => cartItems[0].count += 1);
    const mockRemoveProduct = jest.fn(() => mockProducts.pop());

    render(<ProductList
      isLoading={false}
      products={mockProducts}
      cartItems={cartItems}
      onAddProduct={mockAddProduct}
      onRemoveProduct={mockRemoveProduct}
    />);

    const addButton = screen.getByTestId('AddOutlinedIcon');
    const removeButton = screen.getByTestId('DeleteOutlineIcon');
    fireEvent.click(addButton);
    expect(mockAddProduct).toHaveBeenCalledWith(mockProducts[0]);

    const counter = removeButton.nextSibling;
    expect(counter?.textContent).toBe("1");

    fireEvent.click(removeButton);
    expect(mockRemoveProduct).toHaveBeenCalledWith(mockCartItems[0]);
  });

  it('Deve testar loader', () => {
    render(<ProductList
      isLoading={true}
      products={mockProducts}
      cartItems={[]}
      onAddProduct={() => {}}
      onRemoveProduct={() => {}}
    />);

    const loading = screen.getAllByText('Loading...');
    expect(loading.length).toBe(8);
  });
});
