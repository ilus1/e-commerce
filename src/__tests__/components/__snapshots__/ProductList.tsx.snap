// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente de Listagem de produtos Testa carregamento correto 1`] = `
<ul
  className="sc-kEjbdu MysKI"
  data-testid="product-list"
>
  <li
    className="sc-bdfCDU dsTrPs"
    style={
      {
        "opacity": 0.5,
        "transform": "translateX(30px) translateZ(0)",
      }
    }
  >
    <img
      alt="Produto 1"
      className="sc-gsTDqH bdEbsD"
      src="url1"
    />
    <div
      className="sc-dlfmHC jnmrdn"
    >
      <span
        className="sc-hKgKIp fFuqJh"
      >
        Produto 1
      </span>
      <span
        className="sc-eCstZk iDqxRL"
      >
        R$
        10
      </span>
    </div>
    <span
      className="sc-jSguLX crZTmf"
    >
      Descrição 1
    </span>
    <div
      className="sc-iBPTVF bTHRiV"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-jrAGKZ coCHig css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteOutlineIcon"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
      1
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-fubDmA cxXTFK css-i4bv87-MuiSvgIcon-root"
        data-testid="AddOutlinedIcon"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </div>
  </li>
</ul>
`;
